pynumaflow/mapper/_servicer/_sync_servicer.py

```diff
@@ -0,0 +1,133 @@
+            req = next(request_iterator)
+            # check if it is a valid handshake req
+            if not (req.handshake and req.handshake.sot):
+                raise Exception("MapFn: expected handshake message")
+            yield map_pb2.MapResponse(handshake=map_pb2.Handshake(sot=True))
+
+            # result queue to stream messages from the user code back to the client
+            result_queue = SyncIterator()
+
+            # Reader thread to keep reading from the request iterator and schedule
+            # execution for each of them
+            reader_thread = threading.Thread(
+                target=self._process_requests, args=(context, request_iterator, result_queue)
+            )
+            reader_thread.start()
+            # Read the result queue and keep forwarding them upstream
+            for res in result_queue.read_iterator():
+                # if error handler accordingly
+                if isinstance(res, BaseException):
+                    # Terminate the current server process due to exception
+                    exit_on_error(context, repr(res), parent=self.multiproc)
+                    return
+                # return the result
+                yield res
+
+            # wait for the threads to clean-up
+            reader_thread.join()
+            self.executor.shutdown(cancel_futures=True)
+
+        except BaseException as err:
+            _LOGGER.critical("UDFError, re-raising the error", exc_info=True)
+            # Terminate the current server process due to exception
+            exit_on_error(context, repr(err), parent=self.multiproc)
+            return
+
+    def _process_requests(
+        self,
+        context: NumaflowServicerContext,
+        request_iterator: Iterable[map_pb2.MapRequest],
+        result_queue: SyncIterator,
+    ):
+        try:
+            # read through all incoming requests and submit to the
+            # threadpool for invocation
+            for request in request_iterator:
+                _ = self.executor.submit(self._invoke_map, context, request, result_queue)
+            # wait for all tasks to finish after all requests exhausted
+            self.executor.shutdown(wait=True)
+            # Indicate to the result queue that no more messages left to process
+            result_queue.put(STREAM_EOF)
+        except BaseException as e:
+            _LOGGER.critical("MapFn Error, re-raising the error", exc_info=True)
+            result_queue.put(e)
+
+    def _invoke_map(
+        self,
+        context: NumaflowServicerContext,
+        request: map_pb2.MapRequest,
+        result_queue: SyncIterator,
+    ):
+        try:
+            d = Datum(
+                keys=list(request.request.keys),
+                value=request.request.value,
+                event_time=request.request.event_time.ToDatetime(),
+                watermark=request.request.watermark.ToDatetime(),
+                headers=dict(request.request.headers),
+            )
```

Comment:
Do we need to have this part inside the try except?

Reply:
If there is a type issue, good to catch that mostly